src/Spice86.Core/Emulator/OperatingSystem/DosPathResolver.cs

```diff
@@ -0,0 +1,235 @@
+namespace Spice86.Core.Emulator.OperatingSystem;
+
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+using Spice86.Core.Emulator.OperatingSystem.Structures;
+using Spice86.Shared.Interfaces;
+using Spice86.Shared.Utils;
+
+/// <inheritdoc />
+public class DosPathResolver : IDosPathResolver {
+
+    /// <inheritdoc />
+    public IDictionary<char, MountedFolder> DriveMap { get; private set; } = new Dictionary<char, MountedFolder>();
+
+    /// <inheritdoc />
+    public char CurrentDrive { get; private set; }
+
+    /// <inheritdoc />
+    public string CurrentHostDirectory {
+        get => DriveMap[CurrentDrive].CurrentFolder;
+        private set => DriveMap[CurrentDrive].CurrentFolder = value;
+    }
+
+    private readonly ILoggerService _loggerService;
+
+    /// <summary>
+    /// Initializes a new instance.
+    /// </summary>
+    /// <param name="loggerService">The logger service implementation.</param>
+    public DosPathResolver(ILoggerService loggerService) => _loggerService = loggerService;
+
+    /// <inheritdoc />
+    public string GetHostRelativePathToCurrentDirectory(string path) => Path.GetRelativePath(CurrentHostDirectory, path);
+
+    /// <inheritdoc/>
+    public DosFileOperationResult SetCurrentDir(string dosPath) {
+        if (IsPathRooted(dosPath)) {
+            string? newCurrentDirectory = ToHostCaseSensitiveFullName(dosPath, false);
+            if (!string.IsNullOrWhiteSpace(newCurrentDirectory)) {
+                CurrentHostDirectory = newCurrentDirectory;
+                return DosFileOperationResult.NoValue();
+            }
+        }
+
+        if (dosPath == ".") {
+            return DosFileOperationResult.NoValue();
+        }
+
+        if (dosPath == "..") {
+            CurrentHostDirectory = GetFullNameForParentDirectory(CurrentHostDirectory);
+            return DosFileOperationResult.NoValue();
+        }
+
+        while (dosPath.StartsWith("..\\")) {
+            dosPath = dosPath[3..];
+            CurrentHostDirectory = GetFullNameForParentDirectory(CurrentHostDirectory);
+        }
+
+        CurrentHostDirectory = Path.GetFullPath(Path.Combine(CurrentHostDirectory, dosPath));
+        return DosFileOperationResult.NoValue();
+    }
+
+    /// <inheritdoc/>
+    public void SetDiskParameters(char currentDrive, string dosPath, Dictionary<char, MountedFolder> driveMap) {
+        DriveMap = driveMap;
+        CurrentDrive = currentDrive;
+        SetCurrentDir(dosPath);
+    }
+
+    /// <inheritdoc />
+    public string GetFullNameForParentDirectory(string path) => Directory.GetParent(path)?.FullName ?? path;
+
+    /// <inheritdoc />
+    public string? TryGetFullHostFileName(string dosFilePath) {
+        string? directory = Path.GetDirectoryName(dosFilePath);
```

Comment:
This takes a DOS filepath, but then uses it to search the host filesystem?

Reply:
GetDirectoryName doesn't work with the host file system. It just parses the input.

Reply:
I meant in the foreach a bit later in the code.
If I pass it a DOS path like "D:TOOLS" it will read all the files on my host D drive.

If you look at how it is used you have:
- `Spice86.Core.Emulator.OperatingSystem.DosFileManager.OpenFileInternal()` which passes it a host path.
- `Spice86.Core.Emulator.OperatingSystem.DosPathResolver.ToCaseSensitiveFileName` which passes it a DOS path

They can't both be right.

Reply:
Fixed.